it/controllers/propertyLinking/ClaimPropertyOwnershipISpec.scala

```diff
@@ -0,0 +1,1493 @@
+    Jsoup.parse(res.body)
+  }
+
+  private def stubsSetup(userIsAgent: Boolean, fromCya: Boolean, relationshipChoice: String = "owner", propertyOccupancy: Option[PropertyOccupancy] = None): StubMapping = {
+
+    val relationship: PropertyRelationship = relationshipChoice match {
+      case "owner" => PropertyRelationship(capacity = Owner, uarn = 1)
+      case "occupier" => PropertyRelationship(capacity = Occupier, uarn = 1)
+      case "both" => PropertyRelationship(capacity = OwnerOccupier, uarn = 1)
+    }
+
+    await(
+      mockPropertyLinkingSessionRepository.saveOrUpdate(
+        LinkingSession(
+          address = propertyAddress,
+          uarn = 1L,
+          submissionId = "PL-123456",
+          personId = 1L,
+          earliestStartDate = LocalDate.of(2017, 4, 1),
+          propertyRelationship = Some(relationship),
+          propertyOwnership = if (fromCya) Some(PropertyOwnership(LocalDate.of(yearOfOwnership, monthOfOwnership, dayOfOwnership))) else None,
+          propertyOccupancy = propertyOccupancy,
+          hasRatesBill = None,
+          clientDetails = if (userIsAgent) Some(ClientDetails(123, "Client Name")) else None,
+          localAuthorityReference = localCouncilReferenceValue,
+          rtp = ClaimPropertyReturnToPage.FMBR,
+          fromCya = Some(fromCya),
+          isSubmitted = None
+        )))
+
+    stubFor {
+      get("/business-rates-authorisation/authenticate")
+        .willReturn {
+          aResponse.withStatus(OK).withBody(Json.toJson(testAccounts).toString())
+        }
+    }
+
+    stubFor {
+      post("/auth/authorise")
+        .willReturn {
+          aResponse.withStatus(OK).withBody("{}")
+        }
+    }
+  }
+
+  private def submitOwnership(language: Language = English, day: String, month: String, year: String, userIsAgent: Boolean = false, fromCya: Boolean = true, relationship: String, endDate: Option[PropertyOccupancy] = None): WSResponse = {
+
+    stubsSetup(userIsAgent = userIsAgent, fromCya = fromCya, relationshipChoice = relationship, propertyOccupancy = endDate)
+
+    val requestBody = Json.obj(
+      "interestedStartDate.day" -> day,
+      "interestedStartDate.month" -> month,
+      "interestedStartDate.year" -> year
+    )
+
+    await(
+      ws.url(s"http://localhost:$port/business-rates-property-linking/my-organisation/claim/property-links/ownership")
+        .withCookies(languageCookie(language), getSessionCookie(testSessionId))
+        .withFollowRedirects(follow = false)
+        .withHttpHeaders(HeaderNames.COOKIE -> "sessionId", "Csrf-Token" -> "nocheck")
+        .post(body = requestBody)
+    )
+  }
+
+  private def genericUserPageStuff(day: String, month: String, year: String, relationshipChoice: String, endDate: Option[PropertyOccupancy] = None): Unit = {
+
+    lazy val res = submitOwnership(day = day, month = month, year = year, relationship = relationshipChoice, endDate = endDate)
+
+    lazy val document = Jsoup.parse(res.body)
```

Comment:
I spent ages trying to do something like this before and gave up. My only concern with this way is that we're making a new request each time, even though we just made a request to test the error messages, which could add some extra time to the tests.

I think it would be good to discuss this and hear other people's views on it because I honestly don't know what's best

Reply:
I went through it with Mark, and when i ran the test with it doing the request once/twice, the difference was minimal. (when i ran it with it not doing double requests, on one occasion the tests took longer 😅)